tests/api2/test_api_key.py

```diff
@@ -1,106 +1,237 @@
+
+
+@pytest.mark.parametrize('endpoint', ['LEGACY', 'CURRENT'])
+def test_api_key_session(sharing_admin_user, endpoint):
+    with api_key(sharing_admin_user.username) as key:
         with client(auth=None) as c:
-            assert c.call("auth.login_with_api_key", key)
-            # verify API key works as expected
-            c.call("system.info")
-            # system.version has no_authz_required
-            # this should fail do to lack of authorization for
-            # API key
-            with pytest.raises(Exception):
-                c.call("system.version")
-            with pytest.raises(Exception):
-                c.call("service.start", "cifs")
+            match endpoint:
+                case 'LEGACY':
+                    assert c.call('auth.login_with_api_key', key)
+                case 'CURRENT':
+                    resp = c.call('auth.login_ex', {
+                        'mechanism': 'API_KEY_PLAIN',
+                        'username': sharing_admin_user.username,
+                        'api_key': key
+                    })
+                    assert resp['response_type'] == 'SUCCESS'
+                case _:
+                    raise ValueError(f'{endpoint}: unknown endpoint')
+
+            session = c.call('auth.sessions', [['current', '=', True]], {'get': True})
+            assert session['credentials'] == 'API_KEY'
+            assert session['credentials_data']['api_key']['name'] == 'Test API Key'
+
+            me = c.call('auth.me')
+            assert me['pw_name'] == sharing_admin_user.username
+            assert 'SHARING_ADMIN' in me['privilege']['roles']
+            assert 'API_KEY' in me['account_attributes']
+
+            call("auth.terminate_session", session['id'])
+
             with pytest.raises(Exception):
-                c.call("service.update", "cifs", {"enable": True})
-            auth_token = c.call("auth.generate_token")
+                c.call('system.info')
+
+
+def test_legacy_api_key_upgrade():
+    """We should automatically upgrade old hashes on successful login"""
+    with api_key():
+        key_id = call('api_key.query', [['username', '=', 'root']], {'get': True})['id']
+        call('datastore.update', 'account.api_key', key_id, {
+            'key': LEGACY_ENTRY_HASH,
+            'user_identifier': 'LEGACY_API_KEY'
+        })
+        call('etc.generate', 'pam_middleware')
+
         with client(auth=None) as c:
-            assert c.call("auth.login_with_token", auth_token)
-            # verify that token has same access rights
-            c.call("system.info")
-            with pytest.raises(Exception):
-                c.call("system.version")
-            with pytest.raises(Exception):
-                c.call("service.start", "cifs")
-            with pytest.raises(Exception):
-                c.call("service.update", "cifs", {"enable": True})
+            resp = c.call('auth.login_ex', {
+               'mechanism': 'API_KEY_PLAIN',
```

Comment:
flake8: Cut-n-paste issue.  Every usage of these are under indented
There are seven other of these in this file.

Reply:
I may not be right, but at least I'm consistently wrong.